File: src/main/scala/com/ajjpj/adiagram/geometry/APoint.scala
package com.ajjpj.adiagram.geometry

import scala.language.implicitConversions


/**
 * @author arno
 */
case class APoint(x: Double, y: Double) {
  def halfWayTo(p: APoint) = APoint((x+p.x)/2, (y+p.y)/2)
  def inverse = APoint(-x, -y)

  def +(p: (Double, Double)): APoint = APoint(x+p._1, y+p._2)
  def +(p: APoint): APoint = APoint(x+p.x, y+p.y)
  def +(angle: Angle, d: Double):APoint = this + ((d*angle.unitX, d*angle.unitY))

  def -(p: APoint): APoint = APoint(x-p.x, y-p.y)
}

object APoint {
  val ZERO = APoint(0, 0)

  implicit def pointToPair  (p: APoint) = (p.x, p.y)
  implicit def pointFromPair(p: (Double, Double)) = APoint(p._1, p._2)
}

File: src/main/kotlin/com/ajjpj/adiagram/geometry/APoint.kt
package com.ajjpj.adiagram.geometry

data class APoint(val x: Double, val y: Double) {
    fun halfWayTo(p: APoint) = APoint((x+p.x)/2, (y+p.y)/2)

    val inverse: APoint
        get() = APoint(-x, -y)

    operator fun plus(p: APoint): APoint = APoint(x+p.x, y+p.y)
    fun plus(angle: Angle, d: Double):APoint = this + APoint(d*angle.unitX(), d*angle.unitY())

    operator fun minus(p: APoint): APoint = APoint(x-p.x, y-p.y)

    fun distanceTo(p: APoint) = Math.sqrt((p.x-x)*(p.x-x) + (p.y-y)*(p.y-y))

    companion object {
        val ZERO = APoint(.0, .0)
    }
}